### DISCUSSION TITLE
Proposed shorthand for v-model

### DISCUSSION THREAD

**Discussion by warrensearle** [vuejs/rfcs#395] [RFC Discussions]
# Summary

A shorthand for `v-model` with and without arguments

# Basic example

`::` is shorthand for `v-model`
`::attribute` is shorthand for `v-model:attribute`

# Motivation

Vue provides special shorthands for two of the most often used directives, `v-bind` and `v-on`. [Template shorthands](https://v3.vuejs.org/guide/template-syntax.html#shorthands)

`v-model` is also an often used directive and in Vue 3 it's usage will increase now that we have arguments and a replacement for two-way binding using `.sync`. [Migration guide. v-model](https://v3.vuejs.org/guide/migration/v-model.html#_3-x-syntax)

From a new and regular user point of view `:` and `@` are really nice shorthands for one-way binding and events. Using `::` for two-way binding seems an intuitive partner to `:`.

# Detailed design

**`v-model` Shorthand**
```
<!-- full syntax -->
<ChildComponent v-model="pageTitle" />

<!-- shorthand -->
<ChildComponent ::="pageTitle" />
```

**`v-model` with argument Shorthand**
```
<!-- full syntax -->
<ChildComponent v-model:title="pageTitle" />

<!-- shorthand -->
<ChildComponent ::title="pageTitle" />
```

# Unresolved questions / considerations

- would this have a high implementation cost?
- would the use of `::` be too easily confused with `:`?

**Comment by Demivan** [comment]
Nice idea, but I would assume that `<ChildComponent ::="pageTitle" />` syntax would be disabled.  
Just like you cannot write `<todo-list #="{ item }">` with slot shorthand.

**Comment by stafyniaksacha** [comment]
I like the idea too, but I rather prefer a single character instead  
(why not a `*` star `<ChildComponent *="refValue" *title="refTitle" />`)

**Comment by darrylnoakes** [comment]
Any news on this?

**Comment by sxzz** [comment]
I implemented this feature - `@vue-macros/short-vmodel`. Read the [docs](https://github.com/sxzz/unplugin-vue-macros#shortvmodel).

**Reply by sxzz** [reply]
There are some symbols of prefix that could be considered: `::`, `$`, `*`.

----

Personally, I think `$` is better.

**Reply by darrylnoakes** [reply]
My vote is for `$`.

And I wish GitHub had some sort of "polls" feature...

**Reply by warrensearle** [reply]
I like `$` too. It reminds me of naming variables (e.g. in PHP) and that's kinda what we are setting up here. 

So we would have `<ChildComponent $="pageTitle" />` and `<ChildComponent $title="pageTitle" />`

@sxzz would using `$` work better in terms of the linting issues you experienced with `::` being replaced as `:`?

**Reply by sxzz** [reply]
Yes, at the moment there's better experience using `$` rather than `::`. Cause 3rd party libraries like ESLint, Prettier haven't support it yet.

If Vue official accepted it and merged the RFC, it won't be a problem.

**Reply by sxzz** [reply]
Shorthand is customizable now. There are there prefixes provided: `::`, `$`, `*`.

Example: `<input $="foo" />`

**Reply by yeli19950109** [reply]
https://github.com/vuejs/rfcs/discussions/455#discussion-4304602
```typescript
import { defineConfig } from 'vite'
import vue2 from '@vitejs/plugin-vue2'
import * as compiler from 'vue/compiler-sfc'
import * as cheerio from 'cheerio'
import * as htmlparser2 from 'htmlparser2'
const compileTemplate = compiler.compileTemplate;
import type { Element } from "cheerio"
// https://vitejs.dev/config/
export default defineConfig({
  plugins: [
    vue2({
      compiler: {
        ...compiler,
        compileTemplate: (options) => {
          // @ts-ignore
          const $ = cheerio.load(htmlparser2.parseDocument(options.source, { lowerCaseAttributeNames: false }), { _useHtmlParser2: true })
          const doms = $("*");
          for (let i = 0; i < doms.length; i++) {
            const ele = doms[i] as Element;
            const attrs = ele.attributes.filter(it => it.name.startsWith("$"));
            if (attrs.length) {
              attrs.forEach(it => {
                const el = $(ele).removeAttr(it.name)
                if (it.name == '$') {
                  el.attr("v-model", it.value)
                } else {
                  el.attr(`:${it.name.substring(1, it.name.length)}.sync`, it.value)
                }
              })
            }
          }
          options.source = $.html();
          return compileTemplate(options)
        }
      }
    })
  ],
})
```
vue2中并没有`nodeTransforms`,所以我直接自定义了一个`compileTemplate`,利用`cheerio`直接替换`$`为`v-model`,`$xxx`为`xxx.sync`,然后继续调用`vue/compiler-sfc`自带的compileTemplate

**Comment by jairo-ab** [comment]
Do we have any news for native implementation?

**Comment by angelhdzmultimedia** [comment]
Is this deprecated? 

It's not working for me with either syntax (`$` or `::` )

**Edit:** Ooooh, it's because I'm using Full SFC with JSX. That must be it.

**Edit 2:** Works beautifully so far in .vue files in Nuxt 3!

**Edit 3;** I just realized this is the `vuejs` repo not the `@vue-macros` one. 🤦‍♂️ `shortVModel` is already supported in @vue-macros, just that it only supports `*`, `$` and `::`.

**Comment by wenfangdu** [comment]
@warrensearle How about using `:@`? `:@prop` = `:prop` + `@update:prop`, pretty intuitive.

**Reply by warrensearle** [reply]
Thanks @wenfangdu another good suggestion.

👍 It's using characters we are already familiar with `:` and `@`
👍 It does seem intuitive: something like a 'bind event/update' 😵 
👎 Which way round should they go & should there be any difference between `:@` and `@:`?

**Reply by wenfangdu** [reply]
@warrensearle `@:` should not work, since two options can lead to debates and code inconsistencies.

**Reply by angelhdzmultimedia** [reply]
This was the best option. v-model details in the [docs ](https://vuejs.org/guide/components/v-model.html#basic-usage)explain exactly that v-model combines `  :value="props.modelValue"
    @input="emit('update:modelValue', $event.target.value)"`, so the best syntax for shortVModel would have been `:@` which are the syntax we already using for binding props and events! 😭
I find distasteful how we're using the `$` for everything magic or reactivity related.
    
I hope they consider adding this alternative to choose from.

**Comment by theking2** [comment]
# `#`

`#` It sits next to `@` on most keyhoards. The `$` has too many meanings in JavaScript and other languages so is only confusing. `::` is confusing as one has to distinguish between bind and model. Clearly fine for the experienced dev but not for the junior. `*` is a wildcard. `?` would be an option we are obtaining something from the view and the model. 
But the hastag hints at an identifier. This time for a model parameter, so the semantics are right.

**Comment by petitkriket** [comment]
i would avoid any already used caracter to avoid confusion of responsability with other features

- `#` is already used for slot usage shorthand
- `:`  for props
- `@` for events
- `$` is/was for global variables and comes with history

on the other hand
- `?` does not convey the meaning of two way binding nature of v-model
- `>` is going one way but different that `:`

i am not sure if a v-model shorthand is actually helpful here, not even taking into account nested v-models case

`v-model:streetName="form.streetName"` to `>streetName="form.streetName"`

**Reply by ShGKme** [reply]
`>` is not valid for HTML attribute name

See: https://html.spec.whatwg.org/#attributes-2

**Reply by theking2** [reply]
ofcourse it is not but isn't : or @ not valid either?

**Reply by theking2** [reply]
> i would avoid any already used caracter to avoid confusion of responsability with other features
> 
>     * `#` is already used for slot usage shorthand
> 
>     * `:`  for props
> 
>     * `@` for events
> 
>     * `$` is/was for global variables and comes with history
> 
> 
> on the other hand
> 
>     * `?` does not convey the meaning of two way binding nature of v-model
> 
>     * `>` is going one way but different that `:`
> 
> 
> i am not sure if a v-model shorthand is actually helpful here, not even taking into account nested v-models case
> 
> `v-model:streetName="form.streetName"` to `>streetName="form.streetName"`

I wasn't aware of `#` being used for that already. Thanks! (but is it still used that way🤔)

**Reply by theking2** [reply]
So we are running out of viable ascii characters... `%` `+` `§` `/` `_` ? I suggest we take a vote on what makes sense _and_ is allowed

**Reply by ShGKme** [reply]
> ofcourse it is not but isn't : or @ not valid either?

They are valid. See the link to the specification above.

**Reply by theking2** [reply]
we might want to go in realm of emojis..... 🧩 🤖 . 

Spass bei seite:

char|used
--|--
:|v-bind
#|v-slot
@|v-on
:[]|dynamic v-bind
@[]|dynamic v-on
#[]|dynamic v-slot (? never used it so not sure)

char|why
--|--
`?`|does not show two way reactive
`+`|
`<>`| clashes (optically) with html tags?
`=`|a "wider" `:`
`;`|close enough to `:`
`.`|a single `:` (confuse with class?)
`:@`|kindof v-model plus v-on


```
<!-- full syntax -->
<ChildComponent v-model:title="pageTitle" />

<!-- shorthand -->
<ChildComponent ?title="pageTitle" />
<ChildComponent <>title="pageTitle" />
<ChildComponent =title="pageTitle" />
<ChildComponent $title="pageTitle" />
<ChildComponent &title="pageTitle" />
<ChildComponent ;title="pageTitle" />
<ChildComponent .title="pageTitle" />
<ChildComponent :@title="pageTitle" />

```

Apparently by adding [\poll](https://github.com/apps/polls) installs polls on github

**Reply by angelhdzmultimedia** [reply]
> `:@` 	kindof v-model plus v-on

Can you kindly communicate why do you think `:@` is "**kindof** v-model plus v-on", when to my eyes, it is exactly that? 

Cheers 👋😊

**Reply by wcrowther** [reply]
What is the status of this request?

It is a great idea to have shortened syntax for 2-way binding especially as this have become more mainstream and common using a qualifier. 

For me it is a best practice to provide some intent like v-model.RangeIndex="..." instead of just v-model="...".

My proposal would be to use a percent character like: %RangeIndex="..." as "%" does not have any other VueJs markup context.

**If you think about it - with the diagonal bar of % as the context boundary of a component - then it has symbolic meaning representing a value inside the component and the same value outside the component. This is the gist of 2-way binding!**

To be implemented properly, it should duplicate the v-bind shorthand in Vue 3.4+ where a variable and a prop with the same name can be shortened (like :propName="propName" can be shortened to :propName) -> so %RangeIndex="RangeIndex" could also simply be %RangeIndex.

Users would be encourage to give a meaningful name rather than just %

```
<!-- full syntax -->
<ChildComponent v-model:pageTitle="pageTitle" />

<!-- shorthand -->
<ChildComponent %="pageTitle" />                   <!-- discouraged - use meaningful name -->
<ChildComponent %pageTitle="pageTitle" />          <!-- best practice -->
<ChildComponent %pageTitle  />                     <!-- same-name shorthand similar to param 3.4+ -->

```
